src/target/riscv/riscv-013.c

```diff
@@ -4810,73 +4904,95 @@ static int write_memory_progbuf(struct target *target, target_addr_t address,
-	RISCV013_INFO(info);
-
-	char *progbuf_result = "disabled";
-	char *sysbus_result = "disabled";
-	char *abstract_result = "disabled";
-
-	for (unsigned int i = 0; i < RISCV_NUM_MEM_ACCESS_METHODS; i++) {
-		int method = r->mem_access_methods[i];
-
-		if (method == RISCV_MEM_ACCESS_PROGBUF) {
-			if (mem_should_skip_progbuf(target, address, size, false, &progbuf_result))
-				continue;
-
-			ret = write_memory_progbuf(target, address, size, count, buffer);
+	uint64_t sbver = get_field(info->sbcs, DM_SBCS_SBVERSION);
+	if (sbver == 0)
+		ret = write_memory_bus_v0(target, address, size, count, buffer);
+	else if (sbver == 1)
+		ret = write_memory_bus_v1(target, address, size, count, buffer);
+	else
+		LOG_TARGET_ERROR(target,
+			"Unknown system bus version: %" PRIu64, sbver);
 
-			if (ret != ERROR_OK)
-				progbuf_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_SYSBUS) {
-			if (mem_should_skip_sysbus(target, address, size, 0, false, &sysbus_result))
-				continue;
+	if (ret != ERROR_OK)
+		skip_reason = MEM_ACCESS_STATUS_FAILED;
 
-			if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 0)
-				ret = write_memory_bus_v0(target, address, size, count, buffer);
-			else if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 1)
-				ret = write_memory_bus_v1(target, address, size, count, buffer);
+	return skip_reason;
+}
 
-			if (ret != ERROR_OK)
-				sysbus_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_ABSTRACT) {
-			if (mem_should_skip_abstract(target, address, size, 0, false, &abstract_result))
-				continue;
+static int write_memory(struct target *target, target_addr_t address,
+		uint32_t size, uint32_t count, const uint8_t *buffer)
+{
+	if (!IS_PWR_OF_2(size) || size > 16) {
+		LOG_TARGET_ERROR(target, "BUG: Unsupported size for memory write: %d", size);
+		return ERROR_FAIL;
+	}
 
-			ret = write_memory_abstract(target, address, size, count, buffer);
+	mem_access_status_t skip_reason[] = {
+		[RISCV_MEM_ACCESS_PROGBUF] = MEM_ACCESS_STATUS_DISABLED,
+		[RISCV_MEM_ACCESS_SYSBUS] = MEM_ACCESS_STATUS_DISABLED,
+		[RISCV_MEM_ACCESS_ABSTRACT] = MEM_ACCESS_STATUS_DISABLED
+	};
 
-			if (ret != ERROR_OK)
-				abstract_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_UNSPECIFIED)
-			/* No further mem access method to try. */
-			break;
+	RISCV_INFO(r);
+	for (unsigned int i = 0; i < r->mem_access_method_num; ++i) {
+		riscv_mem_access_method_t method = r->mem_access_methods[i];
+		switch (method) {
+			case RISCV_MEM_ACCESS_PROGBUF:
+				skip_reason[method] =
+					write_memory_progbuf(target, address,
+						size, count, buffer);
+				break;
+			case RISCV_MEM_ACCESS_SYSBUS:
+				skip_reason[method] =
+					write_memory_sysbus(target, address,
+						size, count, buffer);
+				break;
+			case RISCV_MEM_ACCESS_ABSTRACT:
+				skip_reason[method] =
+					write_memory_abstract(target, address,
+						size, count, buffer);
+				break;
+			default:
+				LOG_TARGET_ERROR(target, "Unknown memory access method: %d", method);
```

Comment:
(same as above)

assert(false) can be used here.

```suggestion
				LOG_TARGET_ERROR(target, "Unknown memory access method: %d", method);
				assert(false);
```

Reply:
Addressed